flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2619,157 +2832,541 @@ class ArrayExprLowering {
-          auto tmp = sel.asArray(*e);
+        if (e && !masks->isLowered(e)) {
+          auto optShape =
+              Fortran::evaluate::GetShape(converter.getFoldingContext(), *e);
+          auto tmp = Fortran::lower::createSomeArrayTempValue(
+              converter, optShape, *e, symMap, stmtCtx);
           auto shape = builder.createShape(loc, tmp);
-          masks->vmap.try_emplace(e, fir::getBase(tmp), shape);
+          masks->bind(e, fir::getBase(tmp), shape);
         }
     }
+  }
+
+  /// Build the iteration space into which the array expression will be lowered.
+  /// The resultType is used to create a temporary, if needed.
+  std::pair<IterationSpace, mlir::OpBuilder::InsertPoint>
+  genIterSpace(mlir::Type resultType) {
+    auto loc = getLoc();
+
+    // Generate any mask expressions, as necessary. This is the compute step
+    // that creates the effective masks. See 10.2.3.2 in particular.
+    genMasks();
 
     auto shape = genIterationShape();
-    if (destinationMutableBox) {
-      // Assignment to allocatable array.
-      llvm::SmallVector<mlir::Value> lengthParams;
-      // Currently no safe way to gather length from rhs (at least for
-      // character, it cannot be taken from array_loads since it may be
-      // changed by concatenations).
-      if ((destinationMutableBox->isCharacter() &&
-           !destinationMutableBox->hasNonDeferredLenParams()) ||
-          destinationMutableBox->isDerivedWithLengthParameters())
-        TODO(loc, "gather rhs length parameters in assignment to allocatable");
-
-      llvm::SmallVector<mlir::Value> lbounds;
-      if (takeLboundsIfRealloc && !arrayOperandLoads.empty()) {
-        assert(arrayOperandLoads.size() == 1 &&
-               "lbounds can only come from one array");
-        auto lbs = fir::factory::getOrigins(arrayOperandLoads[0].shape());
-        lbounds.append(lbs.begin(), lbs.end());
-      }
-      Fortran::lower::genReallocIfNeeded(builder, loc, *destinationMutableBox,
-                                         lbounds, shape, lengthParams);
-      // Create ArrayLoad for the rhs and save it into `destination`.
-      PushSemantics(ConstituentSemantics::ProjectedCopyInCopyOut);
-      lowerArrayAssignmentLhs(Fortran::lower::genMutableBoxRead(
-                                  builder, loc, *destinationMutableBox)
-                                  .toExtendedValue());
-      assert(destination && "destination must have been set");
-      // If the rhs is scalar, get shape from the allocatable arrayload.
-      if (shape.empty())
-        shape = getShape(destination);
-    } else if (!destination) {
-      assert(
-          !shape.empty() &&
-          "array expression must have a shape if it has no array destination");
-      // Allocate a storage for the result is it is not already provided.
+    if (!destination) {
+      // Allocate a storage for the result if it is not already provided.
```

Comment:
"Allocate a storage" reads awkwardly.  Did you mean to say "Allocate a storage temp"?